### DISCUSSION TITLE
using multi tenancy database setup with laravel cashier subscriptions with stripe

### DISCUSSION THREAD

**Discussion by MattWiseParking** [spatie/laravel-multitenancy#486] [General]
Has anybody implemented laravel cashier subscriptions with stripe webhooks for a multi tenancy environment.

One method is to create a separate stripe account for each sub domain, but then you would need to add default products/subscriptions upon creating a new tenant and store/use tenant config which you can easily extend the current config().

Another method I have thought about was to use the same account for all sub domains.  you would need to

1. pass in the tenant_id when creating a stripe user.  
2. webhook to check if tenant_id exists against the user, if it does then find the tenant and use the tenant
3. how would we then also change the default Cashier webhooks. i.e stripe/webhook and stripe/payment/{id}, would we need to somehow add in a middleware to these routes, to change the tenant to the correct database prior to executing?

**Comment by winterpk** [comment]
I need this too. Did you ever make any headway here?

**Reply by MattWiseParking** [reply]
I am currently working on integrating this feature in this week.  What it appears that you have to do is in your stripe account create a new account, this will then give you a different set of api secret and key.  What I am then thinking of doing is creating a custom switch tenant task specifically for this.  extending the tenants table with a one to one relationship call the new table something like stripe_api_credentials with fields called  key and secret.  Then in the custom tenant task swap out the config for my current config stripe.key and stripe.secret and replace the values with the relationship fields key/secret.  

Then in the stripe web hook URL,s you can set your domain or sub domain specifically for that database, so when any stripe web hooks make any response it will automatically call the correct database set in your tenants table.

One I have done this I will post my code sample :)

**Comment by MattWiseParking** [comment]
I think there is another way rather than creating a new stripe account for each database.  You will still need a one to one relationship but the table will be for STRIPE_WEBHOOK_SECRET key.  in the stripe dashboard -> webhooks you can create multiple webhook endpoints, each endpoint can represent a separate url for the sub domain.  Once you create a webhook you will then need to retreive the secret (each endpoint will have its own secret)

Next you will then need to create a custom task which will essentially just overwrite config value which i have setup in my system as cashier.webhook.secret which references the environment variable STRIPE_WEBHOOK_SECRET

**Reply by MattWiseParking** [reply]
in your webhook listeners you will need to check that the user exists where the stripe_id = customer variable, if it does then continue with the script.

**Reply by MattWiseParking** [reply]
the only downside to this is that every webhook event will fire to all webhooks (all databases) which is why you must check that the users stripe_id matches the customer variable

**Reply by MattWiseParking** [reply]
cashier.webhook.secret is the default config for cashier
